docs/man/man3/DHT22.3
.TH "DHT22" 3 "Thu May 25 2017" "Version 0.2" "SensorNode" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DHT22 \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <DHT22\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDHT22\fP ()"
.br
.ti -1c
.RI "\fBDHT22\fP (volatile uint8_t &Wire_PORT, volatile uint8_t &Wire_DDR, volatile uint8_t &Wire_PIN, uint8_t Wire_DQ)"
.br
.ti -1c
.RI "int8_t \fBGetTemperatureHumidity\fP (float *temperature, float *humidity)"
.br
.ti -1c
.RI "int8_t \fBGetTemperature\fP (float *temperature)"
.br
.ti -1c
.RI "int8_t \fBGetHumidity\fP (float *humidity)"
.br
.ti -1c
.RI "void \fBGetSensorStringXML\fP (char *string)"
.br
.ti -1c
.RI "void \fBGetSensorTemperatureStringXML\fP (char *string)"
.br
.ti -1c
.RI "void \fBGetSensorHumidityStringXML\fP (char *string)"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBReset\fP ()"
.br
.ti -1c
.RI "uint8_t \fBSendRequest\fP (uint8_t bit)"
.br
.ti -1c
.RI "int8_t \fBGetData\fP (float *temperature, float *humidity)"
.br
.ti -1c
.RI "void \fBAM2302_PIN_INPUT_MODE\fP ()"
.br
.ti -1c
.RI "void \fBAM2302_PIN_OUTPUT_MODE\fP ()"
.br
.ti -1c
.RI "void \fBAM2302_PIN_LOW\fP ()"
.br
.ti -1c
.RI "void \fBAM2302_PIN_HIGH\fP ()"
.br
.ti -1c
.RI "bool \fBAM2302_HIGH_INPUT\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "volatile uint8_t * \fBAM2302_WIRE_PORT\fP"
.br
.ti -1c
.RI "volatile uint8_t * \fBAM2302_WIRE_DDR\fP"
.br
.ti -1c
.RI "volatile uint8_t * \fBAM2302_WIRE_PIN\fP"
.br
.ti -1c
.RI "uint8_t \fBAM2302_WIRE_DQ\fP"
.br
.ti -1c
.RI "uint8_t \fBTIMEOUT\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DHT22::DHT22 ()"

.SS "DHT22::DHT22 (volatile uint8_t & Wire_PORT, volatile uint8_t & Wire_DDR, volatile uint8_t & Wire_PIN, uint8_t Wire_DQ)"

.SH "Member Function Documentation"
.PP 
.SS "bool DHT22::AM2302_HIGH_INPUT ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DHT22::AM2302_PIN_HIGH ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DHT22::AM2302_PIN_INPUT_MODE ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DHT22::AM2302_PIN_LOW ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DHT22::AM2302_PIN_OUTPUT_MODE ()\fC [inline]\fP, \fC [private]\fP"

.SS "int8_t DHT22::GetData (float * temperature, float * humidity)\fC [private]\fP"

.SS "int8_t DHT22::GetHumidity (float * humidity)"

.SS "void DHT22::GetSensorHumidityStringXML (char * string)"

.SS "void DHT22::GetSensorStringXML (char * string)"

.SS "void DHT22::GetSensorTemperatureStringXML (char * string)"

.SS "int8_t DHT22::GetTemperature (float * temperature)"

.SS "int8_t DHT22::GetTemperatureHumidity (float * temperature, float * humidity)"

.SS "void DHT22::Reset ()\fC [private]\fP"

.SS "uint8_t DHT22::SendRequest (uint8_t bit)\fC [private]\fP"

.SH "Member Data Documentation"
.PP 
.SS "volatile uint8_t* DHT22::AM2302_WIRE_DDR\fC [private]\fP"

.SS "uint8_t DHT22::AM2302_WIRE_DQ\fC [private]\fP"

.SS "volatile uint8_t* DHT22::AM2302_WIRE_PIN\fC [private]\fP"

.SS "volatile uint8_t* DHT22::AM2302_WIRE_PORT\fC [private]\fP"

.SS "uint8_t DHT22::TIMEOUT\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for SensorNode from the source code\&.


src/Sensors/DHT22.cpp
/*
 * DHT22.cpp
 *
 *  Created on: Nov 12, 2016
 *      Author: matthias
 */

#include "DHT22.h"
#include <stdio.h>
#include <string.h>
#include <avr/io.h>
#include <util/delay.h>
#include <avr/interrupt.h>
#include <stdlib.h>
#include <ctype.h>

namespace Sensors {

DHT22::DHT22() {


	// PORTD , DDRD , PIND , PD7
	AM2302_WIRE_PORT = &PORTD;
	AM2302_WIRE_DDR = &DDRD;
	AM2302_WIRE_PIN = &PIND;
	AM2302_WIRE_DQ = PD7;
	TIMEOUT=200;



}


DHT22::DHT22(volatile uint8_t &Wire_PORT, volatile uint8_t &Wire_DDR, volatile uint8_t &Wire_PIN,uint8_t Wire_DQ){

	/* Thermometer Connections (At your choice) */
	// PORTD , DDRD , PIND , PD7

	AM2302_WIRE_PORT = &Wire_PORT;
	AM2302_WIRE_DDR = &Wire_DDR;
	AM2302_WIRE_PIN = &Wire_PIN;
	AM2302_WIRE_DQ = Wire_DQ;

	TIMEOUT=200;


}

int8_t DHT22::GetTemperatureHumidity(float *temperature, float *humidity){
	return (GetData(temperature, humidity));
}

int8_t DHT22::GetTemperature(float *temperature){
	float humidity = 0;
	return (GetData(temperature, &humidity));
}

int8_t DHT22::GetHumidity(float *humidity){
	float temperature = 0;
	return (GetData(&temperature, humidity));
}

void DHT22::Reset(){

	//reset port
		AM2302_PIN_OUTPUT_MODE(); //output
		AM2302_PIN_HIGH(); //high
		_delay_ms(100);

}
uint8_t DHT22::SendRequest(uint8_t bit){
	//send request
			AM2302_PIN_LOW(); //low
			_delay_ms(1);
			AM2302_PIN_HIGH(); //high
			AM2302_PIN_INPUT_MODE(); //input
			_delay_us(40);

			//check start condition 1
			if((AM2302_HIGH_INPUT())) {
				return (-1);
			}
			_delay_us(80);
			//check start condition 2
			if(!(AM2302_HIGH_INPUT())) {
				return (-1);
			}
			_delay_us(80);
		return (0);


}
int8_t DHT22::GetData(float *temperature, float *humidity){
	uint8_t bits[5];
	uint8_t i,j = 0;

	memset(bits, 0, sizeof(bits));

		//reset port
		Reset();
		cli();
		//send request
		AM2302_PIN_LOW(); //low
		_delay_ms(1);
		AM2302_PIN_HIGH(); //high
		AM2302_PIN_INPUT_MODE(); //input
		_delay_us(40);

		//check start condition 1
		if((AM2302_HIGH_INPUT())) {
			return (-1);
		}
		_delay_us(80);
		//check start condition 2
		if(!(AM2302_HIGH_INPUT())) {
			return (-2);
		}
		_delay_us(80);

		//read the data
		uint16_t timeoutcounter = 0;
		for (j=0; j<5; j++) { //read 5 byte
			uint8_t result=0;
			for(i=0; i<8; i++) {//read every bit
				timeoutcounter = 0;
				while(!(AM2302_HIGH_INPUT())) { //wait for an high input (non blocking)
					timeoutcounter++;
					if(timeoutcounter > TIMEOUT) {
						return (-3); //timeout
					}
				}
				_delay_us(30);
				if(AM2302_HIGH_INPUT()) //if input is high after 30 us, get result
					result |= (1<<(7-i));
				timeoutcounter = 0;
				while(AM2302_HIGH_INPUT()) { //wait until input get low (non blocking)
					timeoutcounter++;
					if(timeoutcounter > TIMEOUT) {
						return (-4); //timeout
					}
				}
			}
			bits[j] = result;
		}

		//reset port
		Reset();
	sei();
		//check checksum
		if ((uint8_t)(bits[0] + bits[1] + bits[2] + bits[3]) == bits[4]) {
			//return temperature and humidity

			uint16_t rawhumidity = bits[0]<<8 | bits[1];
			uint16_t rawtemperature = bits[2]<<8 | bits[3];
			if(rawtemperature & 0x8000) {
				*temperature = (float)((rawtemperature & 0x7FFF) / 10.0) * -1.0;
			} else {
				*temperature = (float)(rawtemperature)/10.0;
			}
			*humidity = (float)(rawhumidity)/10.0;

			return (0);
		}

		return (-5);




}
void DHT22::GetSensorStringXML(char* string){

	float temp = 0;
	float humidity = 0;
	char tempBuff[20];
	char humBuff[20];
	int m = 0;

	//while(m==-1){
	m = GetTemperatureHumidity(&temp,&humidity);

	//}

	if(m!=0){
		// if an error occurs, port gets reset and interrupt switched on again
		Reset();
		sei();
	}


	dtostrf(temp,3,1,tempBuff);
	dtostrf(humidity,3,1,humBuff);

	sprintf(string,"%s:°C|%s:%%RH|%i|",tempBuff,humBuff,m);


}
void DHT22::GetSensorTemperatureStringXML(char* string){
	float temp = 0;
	float humidity = 0;
	char tempBuff[5];

	GetTemperatureHumidity(&temp,&humidity);
	dtostrf(temp,3,1,tempBuff);


	sprintf(string,"<DHT22><T unit='°C'>%s||",tempBuff);




}

void DHT22::GetSensorHumidityStringXML(char* string){

	float temp = 0;
	float humidity = 0;

	char humBuff[5];
	GetTemperatureHumidity(&temp,&humidity);

	dtostrf(humidity,3,1,humBuff);

	sprintf(string,"<DHT22><H unit='%%RH'>%s||",humBuff);



}

}
